Pawlllosss

Pawlllosss opened the pull request:
It's my final solution. I hope.

bgmarti commented on src/main/java/card/Main.java:
```diff
@@ -0,0 +1,20 @@
+package main.java.card;
```
[NAMING]
Invalid package name, I'd expect something like edu.agh.wfiis.solid.coursework.creditcardvalidator

bgmarti commented on src/main/java/card/information/CreditCardInformation.java:
```diff
@@ -0,0 +1,21 @@
+package main.java.card.information;
```
[NAMING]
Invalid package name, I'd expect something like edu.agh.wfiis.solid.coursework.creditcardvalidator

bgmarti commented on src/main/java/card/issuer/CardIssuerParameters.java:
```diff
@@ -0,0 +1,22 @@
+package main.java.card.issuer;
```
[NAMING]
Invalid package name, I'd expect something like edu.agh.wfiis.solid.coursework.creditcardvalidator

bgmarti commented on src/main/java/card/issuer/CardIssuers.java:
```diff
@@ -0,0 +1,39 @@
+package main.java.card.issuer;
```
[NAMING]
Invalid package name, I'd expect something like edu.agh.wfiis.solid.coursework.creditcardvalidator

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+package main.java.card.validation;
```
[NAMING]
Invalid package name, I'd expect something like edu.agh.wfiis.solid.coursework.creditcardvalidator

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+package main.java.card.validation;
```
[NAMING]
Invalid package name, I'd expect something like edu.agh.wfiis.solid.coursework.creditcardvalidator

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+        //it uses Luhn algorithm
+        ArrayList<Integer> digitsInNumber = stringToDigitsArrayInt(cardNumber);
+
+        int sum = 0 ;
```
[NAMING]
variable names should be descriptive

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+
+        //multiply odd digits by 2
+        for(int i = 0 ; i < digitsInNumber.size() ; i += 2) {
+            int curNmb = digitsInNumber.get(i);
```
[NAMING]
variable names should be descriptive, please do not abbreviate

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+            int curNmb = digitsInNumber.get(i);
+            curNmb *= 2;
+
+            if(curNmb > 9)
```
[READABILITY]
missing parentheses

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+            sum += curNmb;
+        }
+
+        for(int i = 1 ; i < digitsInNumber.size() ; i += 2)
```
[READABILITY]
missing parentheses

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+        for(int i = 1 ; i < digitsInNumber.size() ; i += 2)
+            sum += digitsInNumber.get(i);
+
+        if( sum % 10 == 0)
```
[READABILITY]
missing parentheses

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+
+        if( sum % 10 == 0)
+            return true;
+        else
```
[READABILITY]
this is verbose, could be:
`return sum % 10 == 0`

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+            return false;
+    }
+
+    private static ArrayList<Integer> stringToDigitsArrayInt(String numberInStr){
```
[NAMING]
please do not abbreviate method and variable names

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+    private static ArrayList<Integer> stringToDigitsArrayInt(String numberInStr){
+        ArrayList<Integer> digitsArray = new ArrayList<Integer>();
+
+        for(int i = 0 ; i < numberInStr.length()  ; i++ )
```
[READABILITY]
missing parentheses

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private static ArrayList<Integer> stringToDigitsArrayInt(String numberInStr){
+        ArrayList<Integer> digitsArray = new ArrayList<Integer>();
```
[DESIGN]
should rely on interfaces not implementations

bgmarti replied:
[NAMING]
Array should not be a part of variable name

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+        ArrayList<Integer> lengthsOfCardNumber = cardIssuer.getLengthOfCardNumber();
+
+        for(Integer numberLength : lengthsOfCardNumber)
+            if (numberLength == currentCardNumberLength)
```
[READABILITY]
Missing parentheses

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+        if (!checkCardNumberLength(creditCard))
+            return ResultTypeEnum.ERR_LENGTH;
+
+        if (!validateCardNumber(creditCard.getCardNumber()))
```
[READABILITY]
Missing parentheses

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+        if(!CardIssuers.doesIssuerExist(creditCard.getCardIssuerName()))
+            return ResultTypeEnum.ERR_ISUER;
+
+        if (!checkCardNumberLength(creditCard))
```
[READABILITY]
Missing parentheses

bgmarti commented on src/main/java/card/validation/ValidateCard.java:
```diff
@@ -0,0 +1,103 @@
+
+        CreditCardInformation creditCard = new CreditCardInformation(inputArgs);
+
+        if(!CardIssuers.doesIssuerExist(creditCard.getCardIssuerName()))
```
[READABILITY]
Missing parentheses

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+        return inputArguments[1].matches("[0-9]+");
+    }
+
+    private String [] inputArguments;
```
[READABILITY]
instance variables should be declared above methods

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    private boolean isInputLentgthCorrect(){
+        if( inputArguments.length != expectedArgumentsNmb )
```
[READABILITY]
Missing parentheses

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+    private boolean isInputLentgthCorrect(){
+        if( inputArguments.length != expectedArgumentsNmb )
+            return false;
+        else
```
[READABILITY]
to verbose - can be inlined

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.io.IOException;
+
+public class CardInputArgumentsControl {
```
[NAMING]
invalid class name - does not reflect it's purpose - InputValidator?

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+
+public class CardInputArgumentsControl {
+
+    public CardInputArgumentsControl(String[] inputArgs, int expectedArgsNmb){
```
[NAMING]
do not abbreviate

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+            return true;
+    }
+
+    private boolean isCardNumberCorrect(){
```
[READABILITY]
please see https://docs.oracle.com/javase/7/docs/api/java/lang/CharSequence.html?is-external=true

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public void checkInput() throws IOException{
+        if(!isInputLentgthCorrect())
```
[READABILITY]
Missing parentheses

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+        if(!isInputLentgthCorrect())
+            throw new IOException("Wrong number of input arguments.");
+
+        if(!isCardNumberCorrect())
```
[READABILITY]
Missing parentheses

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+            throw new IOException("Wrong number of input arguments.");
+
+        if(!isCardNumberCorrect())
+            throw new IOException("Didn't provide a card number in a proper format.");
```
[LOGIC]
is this exceptional situation from this method's perspective?

bgmarti commented on src/main/java/card/validation/CardInputArgumentsControl.java:
```diff
@@ -0,0 +1,33 @@
+
+    public void checkInput() throws IOException{
+        if(!isInputLentgthCorrect())
+            throw new IOException("Wrong number of input arguments.");
```
[LOGIC]
is this exceptional situation from this method's perspective?

bgmarti commented on src/main/java/card/Main.java:
```diff
@@ -0,0 +1,20 @@
+
+        ValidateCard.ResultTypeEnum validationResult = ValidateCard.performValidation(args);
+
+        if (validationResult == ValidateCard.ResultTypeEnum.SUCCESS)
```
[READABILITY]
Missing parentheses

bgmarti commented on src/main/java/card/issuer/CardIssuerParameters.java:
```diff
@@ -0,0 +1,22 @@
+
+    public String getCardIssuerName() { return cardIssuerName; }
+
+    private String cardIssuerName;
```
[READABILITY]
instance variables should be declared above methods

bgmarti commented on src/main/java/card/issuer/CardIssuerParameters.java:
```diff
@@ -0,0 +1,22 @@
+    public String getCardIssuerName() { return cardIssuerName; }
+
+    private String cardIssuerName;
+    private ArrayList<Integer> lengthsOfCardNumber;//number of digits for card may be different for one issuer
```
[DESIGN]
should not depend on implementations

bgmarti replied:
[NAMING]
names should be descriptive enough  so that comments are unnecessary

bgmarti commented on src/main/java/card/issuer/CardIssuerParameters.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.util.ArrayList;
+
+//for now it stores only information about length, but it provides possibility to expand program to check other values
```
[DESIGN - OCP]
it does not provides possibility to expand program.  this is not how OCP works

bgmarti replied:
[READABILITY]
any comment's like this should not be added because only clutter the code

bgmarti commented on src/main/java/card/issuer/CardIssuerParameters.java:
```diff
@@ -0,0 +1,22 @@
+//like digits in IIN Range, which identifies the Issuer
+public class CardIssuerParameters {
+
+    CardIssuerParameters(String crdIssuerNm, ArrayList<Integer> lngthOfCrdNmb){
```
[DESIGN] 
should not rely on implementations

bgmarti replied:
[NAMING]
names should be desciptive, any abbreviations are not recommended

bgmarti commented on src/main/java/card/issuer/CardIssuerParameters.java:
```diff
@@ -0,0 +1,22 @@
+
+//for now it stores only information about length, but it provides possibility to expand program to check other values
+//like digits in IIN Range, which identifies the Issuer
+public class CardIssuerParameters {
```
[NAMING]
class name does not reveal it's purpose

bgmarti commented on src/main/java/card/information/CreditCardInformation.java:
```diff
@@ -0,0 +1,21 @@
+        return cardNumber;
+    }
+
+    private String cardIssuerName;
```
[READABILITY]
instance variables should be declared above methods

bgmarti commented on src/main/java/card/information/CreditCardInformation.java:
```diff
@@ -0,0 +1,21 @@
+package main.java.card.information;
+
+
+public class CreditCardInformation {
```
[NAMING]
CreditCardDescription ?

bgmarti commented on src/main/java/card/Main.java:
```diff
@@ -0,0 +1,20 @@
+            System.out.println("Invalid");
+
+        //More precise information about the result
+        System.out.println(validationResult.getMessage());
```
[LOGIC]
pertains only "invalid scenario" isn't it?

bgmarti commented on src/main/java/card/Main.java:
```diff
@@ -0,0 +1,20 @@
+        else
+            System.out.println("Invalid");
+
+        //More precise information about the result
```
[READABILITY]
any comment's like this should not be added because only clutter the code

bgmarti commented on src/main/java/card/information/CreditCardInformation.java:
```diff
@@ -0,0 +1,21 @@
+    }
+
+    private String cardIssuerName;
+    private String cardNumber;//int may be too small, I think it's the better solution than the int array
```
[READABILITY]
any comment's like this should not be added because only clutter the code

bgmarti commented on src/main/java/card/issuer/CardIssuers.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Map;
+
+public class CardIssuers {
+    private static Map<String, CardIssuerParameters> cardIssuersMap =  new HashMap<String, CardIssuerParameters>();
```
[NAMING]
"Map" should not be a part of variable name

bgmarti commented on src/main/java/card/issuer/CardIssuers.java:
```diff
@@ -0,0 +1,39 @@
+    private static Map<String, CardIssuerParameters> cardIssuersMap =  new HashMap<String, CardIssuerParameters>();
+
+    public static boolean doesIssuerExist(String issuerName) {
+        if(cardIssuersMap.get(issuerName) == null)
```
[READABILITY]
Missing parentheses

bgmarti commented on src/main/java/card/issuer/CardIssuers.java:
```diff
@@ -0,0 +1,39 @@
+    public static boolean doesIssuerExist(String issuerName) {
+        if(cardIssuersMap.get(issuerName) == null)
+            return false;
+        else
```
[READABILITY]
code too verbose

bgmarti commented on src/main/java/card/issuer/CardIssuers.java:
```diff
@@ -0,0 +1,39 @@
+
+    public static void prepareCardIssuersMap()
+    {
+        //EXAMPLE DATA(in future it may be created from a file)
```
[READABILITY]
any comment's like this should not be added because only clutters the code

bgmarti commented on src/main/java/card/issuer/CardIssuers.java:
```diff
@@ -0,0 +1,39 @@
+        //EXAMPLE DATA(in future it may be created from a file)
+
+        ArrayList<Integer> digitNumber = new ArrayList<Integer>();
+        //MasterCard
```
any comment's like this should not be added because only clutter the code

bgmarti commented on src/main/java/card/issuer/CardIssuers.java:
```diff
@@ -0,0 +1,39 @@
+    {
+        //EXAMPLE DATA(in future it may be created from a file)
+
+        ArrayList<Integer> digitNumber = new ArrayList<Integer>();
```
[DESIGN]
should not rely on implementations

bgmarti replied:
[NAMING]
name is not descriptive

Pawlllosss commented:
I think I've fixed the biggest violations of SOLID rules. Refactored code. I can't say if it's good right now, but definately better than the last version.

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/issuer/CardIssuersValidators.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class CardIssuersValidators {
```
[NAMING]
this name suggests that validators validate issuer names which is not true - maybe CreditCardValidators ?

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/InputValidator.java:
```diff
@@ -0,0 +1,33 @@
+public class InputValidator {
+
+    private String [] inputArguments;
+    private int expectedArgumentsNmb;
```
[NAMING] 
Nmb - please do not abbreviate

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/InputValidator.java:
```diff
@@ -0,0 +1,33 @@
+        inputArguments = inputArgs;
+    }
+
+    public boolean checkInput(){
```
[LOGIC]
it would be more natural if this method takes input to be validated (String[] inputArgs)  instead of passing it to constructor

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/creditCardValidationController.java:
```diff
@@ -0,0 +1,37 @@
+import edu.agh.wfiis.solid.coursework.creditcardvalidator.card.validation.validator.CreditCardValidator;
+
+
+public class creditCardValidationController {
```
[NAMING]
class name should begin with capital letter, I think that CreditCardsValidator  would be sufficient here and Controller suffix can be omitted

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/creditCardValidationController.java:
```diff
@@ -0,0 +1,37 @@
+        this.inputArguments = inputArguments;
+    }
+
+    public OperationResult performValidation(){
```
[NAMING]
simply validate ?

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/creditCardValidationController.java:
```diff
@@ -0,0 +1,37 @@
+
+        if(!CardIssuersValidators.doesIssuerExist(issuerName)){
+            return new OperationResultErrorIssuerDoesNotExists();
+        }
```
[DESIGN - SRP]
code above should be encapsulated in a separate private method responsible for input checking

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/number/validator/CreditCardNumberValidator.java:
```diff
@@ -0,0 +1,5 @@
+package edu.agh.wfiis.solid.coursework.creditcardvalidator.card.validation.number.validator;
+
+public interface CreditCardNumberValidator {
```
[NAMING]
it is a good practice to name interfaces after their responsibilities using adjectives, ie. ValidatingCardNumberUsingAlgorithm or something better

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/number/validator/CreditCardNumberValidatorLuhn.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class CreditCardNumberValidatorLuhn implements CreditCardNumberValidator {
```
[NAMING]
LuhnValidatingAlgorithm ?

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/number/validator/CreditCardNumberValidatorLuhn.java:
```diff
@@ -0,0 +1,42 @@
+
+        int numberSum = 0;
+
+        //multiply odd digits by 2
```
[READABILITY]
instead of writting comments it is a good practice to encapsulate such code in methods having meaningful name

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/number/validator/CreditCardNumberValidatorLuhn.java:
```diff
@@ -0,0 +1,42 @@
+        return numberSum % 10 == 0;
+    }
+
+    private List<Integer> stringToDigitsArrayInt(String numberInStr){
```
[NAMING]
convertStringToIntegersArray ?

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/result/OperationResult.java:
```diff
@@ -0,0 +1,6 @@
+package edu.agh.wfiis.solid.coursework.creditcardvalidator.card.validation.result;
+
+public interface OperationResult{
```
[NAMING]
ProvidingValidationResult ?

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/Main.java:
```diff
@@ -0,0 +1,23 @@
+package edu.agh.wfiis.solid.coursework.creditcardvalidator.card;
```
[NAMING]
last part (card) does not add value to package name and could be removed - it applies to every class and interface in this project

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/result/OperationResultErrorIssuerDoesNotExists.java:
```diff
@@ -0,0 +1,17 @@
+public class OperationResultErrorIssuerDoesNotExists implements OperationResult{
+
+    private final boolean isSuccess = false;
+    private final String message = "Incorrect issuer name";
```
[READABILITY]
please follow constants naming convention:
`private static final String MESSAGE= "Incorrect issuer name";`

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/result/OperationResultErrorNumberValidation.java:
```diff
@@ -0,0 +1,17 @@
+public class OperationResultErrorNumberValidation implements OperationResult {
+
+    private final boolean isSuccess = false;
+    private final String message = "Card number didn't successfully pass validation";
```
[READABILITY]
please follow constants naming convention

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/result/OperationResultErrorWrongCardNumberLength.java:
```diff
@@ -0,0 +1,17 @@
+public class OperationResultErrorWrongCardNumberLength implements OperationResult{
+
+    private final boolean isSuccess = false;
+    private final String message = "Incorrect card length";
```
[READABILITY]
please follow constants naming convention

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/result/OperationResultErrorWrongInputArguments.java:
```diff
@@ -0,0 +1,17 @@
+public class OperationResultErrorWrongInputArguments implements OperationResult {
+
+    private final boolean isSuccess = false;
+    private final String message = "Didn't provide a proper input";
```
[READABILITY]
please follow constants naming convention

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/result/OperationResultSuccess.java:
```diff
@@ -0,0 +1,17 @@
+public class OperationResultSuccess implements OperationResult{
+
+    private final boolean isSuccess = true;
+    private final String message = "Card validated";
```
[READABILITY]
please follow constants naming convention

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/validator/CreditCardValidator.java:
```diff
@@ -0,0 +1,7 @@
+
+import edu.agh.wfiis.solid.coursework.creditcardvalidator.card.validation.result.OperationResult;
+
+public interface CreditCardValidator {
```
[NAMING]
ValidatingCreditCard

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/validator/CreditCardValidatorMasterCard.java:
```diff
@@ -0,0 +1,37 @@
+import edu.agh.wfiis.solid.coursework.creditcardvalidator.card.validation.result.OperationResultErrorWrongCardNumberLength;
+import edu.agh.wfiis.solid.coursework.creditcardvalidator.card.validation.result.OperationResultSuccess;
+
+public class CreditCardValidatorMasterCard implements CreditCardValidator{
```
[NAMING]
MasterCardValidator?

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/validator/CreditCardValidatorMasterCard.java:
```diff
@@ -0,0 +1,37 @@
+        return numberValidator.validateCardNumber(currentCreditCardNumber);
+    }
+
+    final int creditCardNumberLength = 16;
```
[READABILITY]
it is a constant

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/validator/CreditCardValidatorMasterCard.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    final int creditCardNumberLength = 16;
+    final CreditCardNumberValidator numberValidator = new CreditCardNumberValidatorLuhn();
```
[READABILITY]
variables should be above methods

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/validator/CreditCardValidatorVisa.java:
```diff
@@ -0,0 +1,45 @@
+import edu.agh.wfiis.solid.coursework.creditcardvalidator.card.validation.result.OperationResultErrorWrongCardNumberLength;
+import edu.agh.wfiis.solid.coursework.creditcardvalidator.card.validation.result.OperationResultSuccess;
+
+public class CreditCardValidatorVisa implements CreditCardValidator{
```
[NAMING]
VisaValidator?

bgmarti commented on src/edu/agh/wfiis/solid/coursework/creditcardvalidator/card/validation/validator/CreditCardValidatorVisa.java:
```diff
@@ -0,0 +1,45 @@
+        return numberValidator.validateCardNumber(currentCreditCardNumber);
+    }
+
+    final int [] creditCardNumberLengths = {13, 16, 19};
```
[READABILITY]
please see comments to MasterCard
